src/librustc_data_structures/bitvec.rs

```diff
@@ -15,26 +15,193 @@ pub struct BitVector {
+        for (base, (i, j)) in (a_start..a_end).zip(b_start..b_end).enumerate() {
+            let mut v = self.vector[i] & self.vector[j];
+            for bit in 0..64 {
+                if v == 0 { break; }
+                if v & 0x1 != 0 { result.push(base*64 + bit); }
+                v >>= 1;
+            }
+        }
+        result
+    }
+
+    /// Add the bits from source to the bits from destination,
+    /// return true if anything changed.
+    ///
+    /// This is used when computing reachability because if you have
+    /// an edge `destination -> source`, because in that case
+    /// `destination` can reach everything that `source` can (and
+    /// potentially more).
+    pub fn merge(&mut self, source: usize, destination: usize) -> bool {
+        let (source_start, source_end) = self.range(source);
+        let (destination_start, destination_end) = self.range(destination);
+        let vector = &mut self.vector[..];
+        let mut changed = false;
+        for (source_index, destination_index) in
+            (source_start..source_end).zip(destination_start..destination_end)
+        {
+            let v1 = vector[destination_index];
+            let v2 = v1 | vector[source_index];
+            vector[destination_index] = v2;
+            changed = changed | (v1 != v2);
+        }
+        changed
+    }
+}
+
+fn u64s(elements: usize) -> usize {
+    (elements + 63) / 64
+}
+
+fn word_mask(index: usize) -> (usize, u64) {
+    let word = index / 64;
+    let mask = 1 << (index % 64);
+    (word, mask)
+}
+
+#[test]
+fn union_two_vecs() {
+    let mut vec1 = BitVector::new(65);
+    let mut vec2 = BitVector::new(65);
+    assert!(vec1.insert(3));
+    assert!(!vec1.insert(3));
+    assert!(vec2.insert(5));
+    assert!(vec2.insert(64));
+    assert!(vec1.insert_all(&vec2));
+    assert!(!vec1.insert_all(&vec2));
+    assert!(vec1.contains(3));
+    assert!(!vec1.contains(4));
+    assert!(vec1.contains(5));
+    assert!(!vec1.contains(63));
+    assert!(vec1.contains(64));
+}
+
+#[test]
+fn grow() {
+    let mut vec1 = BitVector::new(65);
+    assert!(vec1.insert(3));
+    assert!(!vec1.insert(3));
+    assert!(vec1.insert(5));
+    assert!(vec1.insert(64));
+    vec1.grow(128);
+    assert!(vec1.contains(3));
+    assert!(vec1.contains(5));
+    assert!(vec1.contains(64));
+    assert!(!vec1.contains(126));
+}
+
+#[test]
+fn matrix_intersection() {
+    let mut vec1 = BitMatrix::new(200);
+
+    vec1.add(2, 3);
+    vec1.add(2, 6);
+    vec1.add(2, 10);
+    vec1.add(2, 64);
+    vec1.add(2, 65);
+    vec1.add(2, 130);
+    vec1.add(2, 160);
+
+    vec1.add(65, 2);
+    vec1.add(65, 8);
+    vec1.add(65, 10); // X
+    vec1.add(65, 64); // X
+    vec1.add(65, 68);
+    vec1.add(65, 133);
+    vec1.add(65, 160); // X
+
+    let intersection = vec1.intersection(2, 64);
+    assert!(intersection.is_empty());
+
+    let intersection = vec1.intersection(2, 65);
+    assert_eq!(intersection, vec![10, 64, 160]);
```

Comment:
nit: the slice `[10, 64, 160]` would suffice here rather than `vec![10, 64, 160]`, right?